device/core/src/client.ts

```diff
@@ -112,10 +128,36 @@ export class Client extends EventEmitter {
     }
 
     this._disconnectHandler = (err) => {
-      this.emit('disconnect', new results.Disconnected(err));
+      debug('transport disconnect event: ' + (err ? err.toString() : 'no error'));
+      if (err && this._retryPolicy.shouldRetry(err)) {
+        if (this._c2dEnabled) {
+          this._c2dEnabled = false;
```

Comment:
Don't have the flow very well in my head yet, but, if did have something like c2dEnabled, and you attempt to do a retry that fails, will you then lose the fact that c2d was originally enabled and on the next retry  you want to start it back up?

Reply:
retries are handled within the _enableC2D call - so retry will work fine until we can't retry and timeout (and at that point, I think we want this to be false.

---
In reply to: [146082947](https://github.com/Azure/azure-iot-sdk-node/pull/114#discussion_r146082947) [](ancestors = 146082947)